### DISCUSSION TITLE
Use of Berry-Script and Scripting together with tasmota

### DISCUSSION THREAD

**Discussion by Sennentuntschi** [arendst/Tasmota#17086] [Q&A]
My Project is to use unused pv-Power for heating.
For this i read the smartmeter values with a script. To control the output i use the DAC of a ESP32. Therefore i use the berry script.
**Now is my problem to get data from the script to the berry-script**. The use of the var1 to 16 is not possible.
Is there a option to do this.
At the moment i use my Iobroker with the the data of the Inverter to control the output of the ESP32.
For the future i hope to integrate this all in the ESP32, standalone unit.

**Comment by s-hadinger** [comment]
Without you sharing the scripts it's difficult to answer. 

You could use a command with `tasmota.cmd()` and parse the result. But I'm not sure there's a command to get a value from scripting. 

Or port everything in Berry

**Reply by Sennentuntschi** [reply]
Everthing in Berry would be nice, but how to read smartmeter with berry?
Here's my script
>D
v1=0
v2=0
>B
->sensor53 r
>M 1
+1,16,s,16,9600,ENERGY
1,77070100100700ff@1,Actual Power,W,Power,0
1,77070100200700ff@1,Voltage L1,V,Voltage,1
1,77070100010800ff@1000,Total consumption,KWh,Total_in,4
1,77070100020800ff@1000,Total Feed,KWh,Total_out,4
1,770701000e0700ff@1,Frequency,Hz,Freq,1
;1,77070100600100ff@#,Server-ID,,ID,0
;1,77070100330700ff@1,Test,,Test,0
;1,7707ffffffffffff@#,Test,,Test,0

>T
v1=ENERGY#Total_out
v2=ENERGY#Power

>S
;every 10 second
if upsecs%10==0
then
print Ausgabe: %v2
;print Schleife %upsecs%
;=>WebSend [192.168.188.110] var10 %v2%
endif

#

**Reply by barbudor** [reply]
If the value you are interested in are part of the SENSOR mqtt message then Berry could subscribe to the sensor topic

**Reply by sfromis** [reply]
If you want to pass values from the script to Berry, you could also execute the `Br command` with a short piece of Berry code to set something. Example command
`Br myobject.setValue(123, 456)`
Of course, you need the usual Scripting language prefix for the command too, `=>` or whatever....

This is basically quite similar to using the WebSend command in the code you posted

**Reply by Sennentuntschi** [reply]
> If the value you are interested in are part of the SENSOR mqtt message then Berry could subscribe to the sensor topic

With **tasmota.read_sensors()** it's possible to get values into berry
Just checked with this 
def t() print(**tasmota.read_sensors()**) tasmota.set_timer(2000, t) end
tasmota.set_timer(2000, t)

Thanks

**Comment by Sennentuntschi** [comment]
Found other options to get values from smartmeter, 
def rule_Power(value)
  print("Power: ",value)
end
tasmota.add_rule("ENERGY#Power",rule_Power)

New values depend on settings in logging

For faster updates use this
import json
class energy_driver
 def every_second()
 sensors=json.load(tasmota.read_sensors())
 print(sensors['ENERGY']['Power'])
 end
end
tasmota.add_driver(energy_driver())

**Reply by Sfre12** [reply]
Hi Sennentuntschi.
I am very interested in your solution.
Right now, I have a ESP8266 with SML running to read the smart meter.
But I want to change to ESP32 and berry language to have more possibilities and also use the analog output of the ESP32.

Can you please explain to me in more detail how you now read the smart meter with the SML language and then process it further in Berry? The script(s) for this would be very helpful.
Or have you now read and processed the smart meter completely in Berry?
I see you have some German words in your code, you can send it also in German by PN
I am from Germany

**Reply by Sennentuntschi** [reply]
Hi Sfre12,
it took  me a view nights, but in the end it's easy.
The smartmeter values are read with the normal script , how you find it in the smartmeter description.
Berrys just prints the value at the console.
At the moment the iobroker does the job to look wether or not there is energy left for heating.
But I'm thinking about a standalone version without smarthome.

Here ist my script for the efr SGM-C4-4A920L
>D
>B
->sensor53 r
>M 1
+1,16,s,16,9600,ENERGY
1,77070100100700ff@1,Actual Power,W,Power,0
1,77070100200700ff@1,Voltage L1,V,Voltage,1
1,77070100010800ff@1000,Total consumption,KWh,Total_in,4
1,77070100020800ff@1000,Total Feed,KWh,Total_out,4
1,770701000e0700ff@1,Frequency,Hz,Freq,1
;1,77070100600100ff@#,Server-ID,,ID,0
;1,77070100330700ff@1,Test,,Test,0
#

`The berry script `
`var bOn
var iPower
import json

class energy_driver


 def stop_activ()
  bOn = false  
 end

 def start_activ()
  bOn = true 
 end

 def every_second()
   var sensors=json.load(tasmota.read_sensors())
   iPower=sensors['ENERGY']['Power']
   if bOn==true print("Grid: ", iPower) end
   
 end
end

tasmota.add_driver(energy_driver()) `

**Reply by Sfre12** [reply]
Thanks a lot to you. I think this will help.
Because I am at the very beginning for berry scripting I have some more questions :-)
1. As I understood is this berry script getting "Power", "Voltage", "Total_in", "Total_out", "Freq", "ID" and  "Test" to a variable "sensors" and then you pick only "Power" with the next line (iPower=sensors['ENERGY']['Power']). 
So I can use the same to extract the other parameters, like: "iVoltage=sensors['ENERGY']['Voltage']" ?
2. Why you add a driver for this and not running the script every cycle?
3. With line (var sensors=json.load(tasmota.read_sensors())) your are grabbing ALL sensors? (In case there is also a DS18B20 connected to the device?

**Reply by sfromis** [reply]
Yes, `tasmota.read_sensors()` gets the same JSON string as you see at TelePeriod published to MQTT, and also printed in the console. You have all the same sensors available.

`json.load()` parses it into the map structure `sensors` containing multiple key/value pairs, in a hierarchy closely matching what you see in the JSON.

Where you add it depends on when you want to use the values. `every_second` was just a simple example, but you could also have a `cron` timer to do it at regular times. The point of adding it a driver is to have `every_second()`, but as mentioned you can also manage the calling frequency different ways. Depending on what you mean by "every cycle", you could do that.

You can indeed refer to values the way you suggest. You could also do things like:
```
var sensors = json.load(tasmota.read_sensors())
var energy = sensors['ENERGY]
iPower = energy['Power']
iVoltage = energy['Voltage']
```
This takes the energy map into a separate variable, instead of repeated references to the sensors variable.

**Reply by Sfre12** [reply]
Thanks!!!! This is very helpfull

**Reply by Sennentuntschi** [reply]
Good to know, thanks!
